### PR TITLE
[MLIR][Linalg] Ternary Op & Linalg select

### PR DESCRIPTION
**Pull request by kurapov-peter**
Following #90236, adding `select` to linalg as `arith.select`. No implicit type casting.
OpDSL doesn't expose a type restriction for bool, but I saw no reason in adding it (put a separate symbolic type and check the semantics in the builder).

### CODE REVIEW

File: mlir/python/mlir/dialects/linalg/opdsl/lang/comprehension.py [hunk 1]
```diff
@@ -339,6 +340,30 @@ class BinaryFn:
     powf = BinaryFnType("powf")
 
 
+class TernaryFnType:
+    """Ternary function.
+
+    A bterary function takes three tensor expressions and returns the
```
Comment (makslevental) [inline mlir/python/mlir/dialects/linalg/opdsl/lang/comprehension.py:346]:
```suggestion
    A ternary function takes three tensor expressions and returns the
```

File: mlir/test/Dialect/Linalg/named-ops.mlir [hunk 2]
```diff
@@ -1924,3 +1924,25 @@ func.func @fill_tensor(%f: f32, %v: vector<2x4xf32>) -> (tensor<f32>, tensor<vec
...
+  // CHECK-SAME: ins(%{{.+}}, %{{.+}}, %{{.+}} : memref<?x?x?xi1>, memref<?x?x?xf32>, memref<?x?x?xf32>)
+  // CHECK-SAME: outs(%{{.+}} : memref<?x?x?xf32>)
+  linalg.select ins(%arg0, %arg1, %arg2 : memref<?x?x?xi1>, memref<?x?x?xf32>, memref<?x?x?xf32>) outs(%arg3: memref<?x?x?xf32>)
+  return
+}
+
+// -----
+
+// CHECK-LABEL: func @select_static
+func.func @select_static(%arg0: memref<4x8x16xi1>, %arg1: memref<4x8x16xf32>, %arg2: memref<4x8x16xf32>, %arg3: memref<4x8x16xf32>) {
```
Comment (ciyongch) [inline mlir/test/Dialect/Linalg/named-ops.mlir:1942]: Better to cover `tensor` besides `memref`?

Comment (rengolin) [reply mlir/test/Dialect/Linalg/named-ops.mlir:1942]: Ah, yes. The other ops have a tensor test, too, just copy and paste from them.

**Comment by rengolin** [comment]
ping

**Review by matthias-springer — APPROVED** [review]
I'm not too familiar with the Op DSL part, but this op makes sense to me.

There is actually a TODO in the `BufferizableOpInterface` implementation of `arith.select`: only `i1` conditions are supported. That is because there is no efficient way to bufferize `arith.select` with tensor conditions because the op is not in destination style. It is good to have a `linalg.select` destination-style op that can bufferize efficiently.

**Pull request merged**